geode-core/src/main/java/org/apache/geode/internal/cache/FilterProfile.java

```diff
@@ -1846,18 +1846,27 @@ protected void process(ClusterDistributionManager dm) {
         }
 
         CacheDistributionAdvisor cda = (CacheDistributionAdvisor) r.getDistributionAdvisor();
-        CacheDistributionAdvisor.CacheProfile cp =
-            (CacheDistributionAdvisor.CacheProfile) cda.getProfile(getSender());
-        if (cp == null) { // PR accessors do not keep filter profiles around
-          if (logger.isDebugEnabled()) {
-            logger.debug(
-                "No cache profile to update, adding filter profile message to queue. Message :{}",
-                this);
+        CacheDistributionAdvisor.CacheProfile cp;
+
+        // prevent adding the message to queue after we have processed the queue
+        // in CreateRegionReplyProcessor.process
+        synchronized (cda) {
```

Comment:
It is not clear to me how `synchronize (cda)` can achieve your solution:
>When we are at step 5, synchronize on the advisor so that 5 finishes before 6 or completely after 6.

I don't see `CacheDistributionAdvisor` is synchronized on `CreateRegionReplyProcessor`. Also my understanding is that on the receiver side, there is only one thread that processes this single instance of `OperationMessage`. Maybe I am missing something. Can you tell me what are the threads that compete for the lock of `cda` here?
And for the synchronized block, why it does not include the case when cp != null?

Reply:
It's synchronized on line 249 in the `CreateRegionReplyProcessor.process()` method, the`putProfile` is a synchronized method on the `cda` object itself. So either that finishes first or this block finishes first. The lock only needs to be retained in the `if` block of the code where we are adding the message to the queue, for `else` block, we don't need to retain the lock.